File: apps/web/src/offline/memory-sync-repository.test.ts
import assert from "node:assert/strict";
import test from "node:test";
import { MemorySyncRepository } from "./memory-sync-repository.ts";
import {
  SYNC_PROTOCOL_VERSION,
  type SyncCommandEnvelope,
} from "./sync-contract.ts";

const partition = { tenantId: "tenant-a", deviceId: "device-1" } as const;
const commandId = "11111111-1111-4111-8111-111111111111";
const command = (
  id: string,
  payload = { total: 100 },
): SyncCommandEnvelope => ({
  version: SYNC_PROTOCOL_VERSION,
  commandId: id,
  type: "sales.finalize.v1",
  occurredAt: "2026-08-08T10:00:00.000Z",
  payload,
});

test("outbox deduplicates exact commands and rejects reused ids with changed payload", async () => {
  const repository = new MemorySyncRepository();
  await repository.enqueue(partition, command(commandId));
  await repository.enqueue(partition, command(commandId));
  assert.equal((await repository.outbox(partition)).length, 1);
  await assert.rejects(
    repository.enqueue(partition, command(commandId, { total: 101 })),
    /different envelope/,
  );
});

test("claim, interruption recovery, retry, and terminal state are durable state transitions", async () => {
  const repository = new MemorySyncRepository();
  const now = new Date("2026-08-08T10:01:00.000Z");
  await repository.enqueue(partition, command(commandId), now);
  const [first] = await repository.claimPending(partition, 10, now);
  assert.equal(first.state, "sending");
  assert.equal(first.attempts, 1);
  assert.equal(await repository.recoverInterrupted(partition, now), 1);
  const [retry] = await repository.claimPending(partition, 10, now);
  assert.equal(retry.attempts, 2);
  await repository.settle(
    partition,
    commandId,
    "conflict",
    "inventory_changed",
  );
  assert.deepEqual((await repository.outbox(partition))[0], {
    ...retry,
    state: "conflict",
    failureCode: "inventory_changed",
  });
});

test("projection batches apply atomically with a monotonic cursor", async () => {
  const repository = new MemorySyncRepository();
  await repository.applyChanges(partition, {
    version: SYNC_PROTOCOL_VERSION,
    cursor: 2,
    hasMore: false,
    changes: [
      {
        cursor: 1,
        entityType: "products",
        entityId: "b",
        operation: "upsert",
        payload: { name: "Bread" },
        occurredAt: "2026-08-08T10:00:00Z",
      },
      {
        cursor: 2,
        entityType: "products",
        entityId: "a",
        operation: "upsert",
        payload: { name: "Milk" },
        occurredAt: "2026-08-08T10:00:01Z",
      },
    ],
  });
  assert.equal(await repository.cursor(partition), 2);
  assert.deepEqual(
    (await repository.projection(partition, "products")).map(
      (item) => item.entityId,
    ),
    ["a", "b"],
  );

  await assert.rejects(
    repository.applyChanges(partition, {
      version: SYNC_PROTOCOL_VERSION,
      cursor: 1,
      hasMore: false,
      changes: [],
    }),
    /backwards/,
  );
  assert.equal(await repository.cursor(partition), 2);
  assert.equal((await repository.projection(partition, "products")).length, 2);
});

test("tenant and device partitions do not share cursors, projections, or outbox rows", async () => {
  const repository = new MemorySyncRepository();
  await repository.enqueue(partition, command(commandId));
  await repository.applyChanges(partition, {
    version: SYNC_PROTOCOL_VERSION,
    cursor: 1,
    hasMore: false,
    changes: [
      {
        cursor: 1,
        entityType: "products",
        entityId: "a",
        operation: "upsert",
        payload: { name: "Milk" },
        occurredAt: "2026-08-08T10:00:00Z",
      },
    ],
  });
  const other = { tenantId: "tenant-a", deviceId: "device-2" };
  assert.equal(await repository.cursor(other), 0);
  assert.deepEqual(await repository.projection(other, "products"), []);
  assert.deepEqual(await repository.outbox(other), []);
});

test("schema and malformed change batches fail closed before mutation", async () => {
  const repository = new MemorySyncRepository();
  await assert.rejects(
    repository.enqueue(partition, {
      ...command(commandId),
      version: "2" as "1",
    }),
    /Unsupported/,
  );
  await assert.rejects(
    repository.applyChanges(partition, {
      version: SYNC_PROTOCOL_VERSION,
      cursor: 3,
      hasMore: false,
      changes: [
        {
          cursor: 2,
          entityType: "products",
          entityId: "a",
          operation: "upsert",
          payload: {},
          occurredAt: "2026-08-08T10:00:00Z",
        },
        {
          cursor: 1,
          entityType: "products",
          entityId: "b",
          operation: "delete",
          payload: {},
          occurredAt: "2026-08-08T10:00:01Z",
        },
      ],
    }),
    /increasing/,
  );
  assert.equal(await repository.cursor(partition), 0);
});

test("partition reset clears only the selected tenant and device", async () => {
  const repository = new MemorySyncRepository();
  const other = { tenantId: "tenant-a", deviceId: "device-2" } as const;
  await repository.enqueue(partition, command(commandId));
  await repository.enqueue(
    other,
    command("22222222-2222-4222-8222-222222222222"),
  );
  await repository.applyChanges(partition, {
    version: SYNC_PROTOCOL_VERSION,
    cursor: 4,
    hasMore: false,
    changes: [
      {
        cursor: 4,
        entityType: "products",
        entityId: "a",
        operation: "upsert",
        payload: { name: "Milk" },
        occurredAt: "2026-08-08T10:00:00Z",
      },
    ],
  });

  await repository.resetPartition(partition);

  assert.equal(await repository.cursor(partition), 0);
  assert.deepEqual(await repository.outbox(partition), []);
  assert.deepEqual(await repository.projection(partition, "products"), []);
  assert.equal((await repository.outbox(other)).length, 1);
});

File: apps/mobile/test/memory_sync_repository_test.dart
import 'package:nova_mobile/nova_mobile.dart';
import 'package:test/test.dart';

void main() {
  const a = SyncPartition(tenantId: 'tenant-a', deviceId: 'device-a');
  const b = SyncPartition(tenantId: 'tenant-b', deviceId: 'device-a');

  OutboxCommand command(Map<String, Object?> payload) => OutboxCommand(
        commandId: '018f0000-0000-7000-8000-000000000001',
        type: 'sales.finalize.v1',
        payload: payload,
        occurredAt: DateTime.utc(2026, 8, 8),
      );

  test('partitions cursors and projections by tenant and device', () async {
    final repository = MemorySyncRepository();
    await repository.applyChanges(a, 0, [
      ProjectionChange(
        cursor: 1,
        entityType: 'product',
        entityId: 'p1',
        operation: 'upsert',
        payload: {'name': 'Milk'},
        occurredAt: DateTime.utc(2026, 8, 8),
      ),
    ]);

    expect(await repository.cursor(a), 1);
    expect(await repository.cursor(b), 0);
    expect(await repository.projection(b, 'product', 'p1'), isNull);
  });

  test('accepts gaps and rejects a reordered batch atomically', () async {
    final repository = MemorySyncRepository();
    await repository.applyChanges(a, 0, [
      ProjectionChange(
        cursor: 2,
        entityType: 'product',
        entityId: 'p0',
        operation: 'upsert',
        payload: {'name': 'Flour'},
        occurredAt: DateTime.utc(2026, 8, 8),
      ),
    ]);
    await expectLater(
      repository.applyChanges(a, 2, [
        ProjectionChange(
          cursor: 5,
          entityType: 'product',
          entityId: 'p1',
          operation: 'upsert',
          payload: {'name': 'Milk'},
          occurredAt: DateTime.utc(2026, 8, 8),
        ),
        ProjectionChange(
          cursor: 4,
          entityType: 'product',
          entityId: 'p2',
          operation: 'upsert',
          payload: {'name': 'Bread'},
          occurredAt: DateTime.utc(2026, 8, 8),
        ),
      ]),
      throwsA(isA<SyncRepositoryException>()),
    );
    expect(await repository.cursor(a), 2);
    expect(await repository.projection(a, 'product', 'p1'), isNull);
  });

  test('deduplicates commands and rejects changed replay payloads', () async {
    final repository = MemorySyncRepository();
    await repository.enqueue(a, command({'sale': 'one', 'total': 100}));
    await repository.enqueue(a, command({'total': 100, 'sale': 'one'}));
    expect(await repository.pending(a), hasLength(1));
    await expectLater(
      repository.enqueue(a, command({'sale': 'two', 'total': 100})),
      throwsA(isA<SyncRepositoryException>()),
    );
  });

  test('recovers interrupted sends for retry', () async {
    final repository = MemorySyncRepository();
    await repository.enqueue(a, command({'sale': 'one'}));
    await repository.markSending(a, command({}).commandId);
    expect(await repository.pending(a), isEmpty);
    await repository.recoverInterrupted(a);
    final retried = await repository.pending(a);
    expect(retried.single.status, OutboxStatus.pending);
    expect(retried.single.attempts, 1);
  });
}
